### ISSUE TITLE
Bootstraping ignores make jobserver

### ISSUE THREAD

**Issue by jefferyto** [rust-lang/rust#116101]
I'm currently working on enhancing the [Rust package](https://github.com/openwrt/packages/blob/master/lang/rust/Makefile) (currently 1.72.0) in OpenWrt (the build system is based on buildroot). After ensuring the x.py command is prefixed with `+` and that `MAKEFLAGS` is set correctly, bootstraping continues to use all of my CPUs instead of the `-jN` jobs from the make command line.

I have tried patching out [these lines](https://github.com/rust-lang/rust/blob/1.72.0/src/bootstrap/builder.rs#L1217-L1219) that remove `MAKEFLAGS` from the environment for cargo; the `MAKEFLAGS` value does reach cargo but cargo continues to set up its own jobserver instead.

I can see that removing `MAKEFLAGS` from the environment was deliberate (as noted in https://github.com/rust-lang/rust/pull/50629#issuecomment-389683073), but it would be great if building rustc can cooperate with the rest of the build system by using the same jobserver.

**Comment by onur-ozkan** [comment]
> I can see that removing MAKEFLAGS from the environment was deliberate (as noted in https://github.com/rust-lang/rust/pull/50629#issuecomment-389683073), but it would be great if building rustc can cooperate with the rest of the build system by using the same jobserver.

Unfortunately, achieving this 100% during the bootstrapping process is not possible. However, there is a workaround provided in this PR: https://github.com/rust-lang/rust/pull/116108, which allows you to set the cargo jobs to the value provided to Make using `-jN`, similar to how it's done with `x.py -jN`.

> bootstraping continues to use all of my CPUs instead of the -jN jobs from the make command line.

The PR linked above should resolve this problem.

**Comment by jefferyto** [comment]
> However, there is a workaround provided in this PR: https://github.com/rust-lang/rust/pull/116108, which allows you to set the cargo jobs to the value provided to Make using `-jN`, similar to how it's done with x.py `-jN`.

Thanks but that doesn't solve the issue. When I do a build in OpenWrt, I am often building many packages, not just Rust. If we just give the same `-jN` value to Rust, say `-j4`, then the overall build system will think there are 4 jobs available and Rust will also think there are 4 jobs available. Then potentially the overall build system will be building 3 other packages and Rust will be using 4 jobs, i.e. total of 7 jobs. This is why Rust should use the same jobserver as make.

**Comment by onur-ozkan** [comment]
I thought you were only building rust from make, I see now. Thanks for clarification.

> I have tried patching out [these lines](https://github.com/rust-lang/rust/blob/1.72.0/src/bootstrap/builder.rs#L1217-L1219) that remove MAKEFLAGS from the environment for cargo; the MAKEFLAGS value does reach cargo but cargo continues to set up its own jobserver instead.

I assume the patch was removing those lines, right?

**Comment by jefferyto** [comment]
> I assume the patch was removing those lines, right?

Yes. I removed the lines from `bare_cargo()` not `rustdoc_cmd()` though (the [commit](https://github.com/rust-lang/rust/commit/209240dc267075d69eb8d1dba23be2a8c80c6427) you referenced in https://github.com/rust-lang/rust/pull/116108#issuecomment-1732519448 was for `rustdoc_cmd()`); I assume `rustdoc_cmd()` is only invoked when building rustdoc.

**Comment by onur-ozkan** [comment]
Then this

> the MAKEFLAGS value does reach cargo but cargo continues to set up its own jobserver instead.

means the real issue is more related with cargo itself.

**Comment by onur-ozkan** [comment]
We can remove

https://github.com/rust-lang/rust/blob/70a7fe174321284ab262b64941b123ad081af242/src/bootstrap/builder.rs#L1246-L1248

lines once 

> the MAKEFLAGS value does reach cargo but cargo continues to set up its own jobserver instead.

is fixed from cargo side.

**Comment by jefferyto** [comment]
> means the real issue is more related with cargo itself.

I'm able to have cargo use the same make jobserver (to build other Rust packages) after rustc is compiled/installed (cargo is called directly from the Makefile), so I would say that bootstrap is at least involved in this issue, if not the main cause.
